### ISSUE TITLE
problems with xilinx_emacps.c and macb.c in xilinx-v2015.3

### ISSUE THREAD

**Issue by pavel-demin** [Xilinx/linux-xlnx#67]
Hello,

I'm trying to upgrade my Linux kernel from xilinx-v2015.2 to xilinx-v2015.3 and I've found a couple of problems with [drivers/net/ethernet/xilinx/xilinx_emacps.c](https://github.com/Xilinx/linux-xlnx/blob/xilinx-v2015.3/drivers/net/ethernet/xilinx/xilinx_emacps.c) and [drivers/net/ethernet/cadence/macb.c](https://github.com/Xilinx/linux-xlnx/blob/xilinx-v2015.3/drivers/net/ethernet/cadence/macb.c):
 - I had to add ```#include <linux/timecounter.h>```, otherwise the build failed with ```error: implicit declaration of function 'timecounter_init'```,
 - during the shutdown, the ```macb``` driver crashes at [line 1708 ](https://github.com/Xilinx/linux-xlnx/blob/xilinx-v2015.3/drivers/net/ethernet/cadence/macb.c#L1708) when calling ```ptp_clock_unregister(bp->ptp_clock)```.

I'm testing it on the Red Pitaya board.

Best regards,

Pavel

**Comment by michalsimek** [comment]
Hi,

any particular reason why you want to use emacps driver? As you can see all boards switched to macb driver which is tested. The only one reason why there is emacps driver is that there is gmii2rgmii bridge which we are going to support in macb in future release.

Thanks,
Michal

**Comment by michalsimek** [comment]
Hi,

any reason to enable CONFIG_XILINX_PS_EMAC_HWTSTAMP=y? 

Thanks,
Michal

**Comment by pavel-demin** [comment]
Hi Michal,

The driver crashes even without ```CONFIG_XILINX_PS_EMAC_HWTSTAMP=y```.

Enabling CONFIG_XILINX_PS_EMAC_HWTSTAMP is suggested in the following article:
http://www.wiki.xilinx.com/Zynq-7000+AP+SoC+-+Precision+Timing+with+IEEE1588+v2+Protocol+Tech+Tip

With ```CONFIG_XILINX_PS_EMAC=y``` removed from ```arch/arm/configs/xilinx_zynq_defconfig```, the problem still persists.

With ```CONFIG_MACB=y``` removed from ```arch/arm/configs/xilinx_zynq_defconfig```, no more crashes during the shutdown.

Looks like I was looking in the wrong place and it's ```ptp_clock_unregister(bp->ptp_clock)``` on  [line 1708 in macb.c](https://github.com/Xilinx/linux-xlnx/blob/xilinx-v2015.3/drivers/net/ethernet/cadence/macb.c#L1708) that crashes.

Regards,

Pavel

**Comment by pavel-demin** [comment]
As a workaround, I've taken the previous version of [macb.h](https://raw.githubusercontent.com/Xilinx/linux-xlnx/f3ad0c4ffd3f20dc0604a9c805768d818ff839df/drivers/net/ethernet/cadence/macb.h) and [macb.c](https://raw.githubusercontent.com/Xilinx/linux-xlnx/f3ad0c4ffd3f20dc0604a9c805768d818ff839df/drivers/net/ethernet/cadence/macb.c). The previous version does not contain any PTP code and the driver does not crash during shutdown.

To summarize:
 - current version (https://github.com/Xilinx/linux-xlnx/commit/8fddb0b6216d174a750424440944cad639b80247) of the macb driver crashes during shutdown,
 - previous version (https://github.com/Xilinx/linux-xlnx/commit/f3ad0c4ffd3f20dc0604a9c805768d818ff839df) of the macb driver works without any problem.

**Comment by pavel-demin** [comment]
I've just noticed the latest commit (https://github.com/Xilinx/linux-xlnx/commit/f487868d3f26ed588e7bd55f8af73e29316a47dc) that fixes the problem with the macb driver.

Thanks a lot for fixing this so quickly!

Looking at this commit and at the macb code, I've realized that the new PTP functionality only works when compatible is set to "cdns,zynqmp-gem". So, I have a slightly off-topic question. Is this new PTP functionality in the macb driver compatible with the Zynq-7000 devices?

**Comment by michalsimek** [comment]
FW:
This new PTP functionality was only added in Zynq Ultrascale+ MPSoC (cdns, zynqmp-gem).
It is not the same as Zynq 7000 (cdns,gem or cdns,zynq-gem). That was implemented in xilinx_emacps.c and has not yet been merged to macb.c.

Thanks,
Harini

**Comment by pavel-demin** [comment]
Just tested the latest version of the macb driver. It works without any problem now.

Closing this issue.
